File: designpattern/designpattern.go
package main

import (
	"fmt"
	"sync"
)

/* -------------------------------------------------------------------------- */
/*                                 single instace                             */
/* -------------------------------------------------------------------------- */

type singleton struct {
	timestamp int64
}

var instance *singleton
var once sync.Once

func GetInstance() *singleton {
	once.Do(func() {
		instance = &singleton{timestamp: 123456789}
	})
	return instance
}

// func main() {
// 	a := GetInstance()
// 	b := GetInstance()
// 	fmt.Println(a == b) // true
// }

/* -------------------------------------------------------------------------- */
/*                                 factory pattern                            */
/* -------------------------------------------------------------------------- */

type Vehicle interface { Drive() }

type Car struct{}
func (c Car) Drive() { fmt.Println("Driving a car") }

type Truck struct{}
func (t Truck) Drive() { fmt.Println("Driving a truck") }

func VehicleFactory(t string) Vehicle {
    if t == "car" { return Car{} }
    if t == "truck" { return Truck{} }
    return nil
}

func main() {
    v := VehicleFactory("car")
    v.Drive()
}

File: designpattern/designpattern.js
/* -------------------------------------------------------------------------- */
/*                                 single instace                             */
/* -------------------------------------------------------------------------- */

class Singleton {
    constructor() {
        if (Singleton.instance) return Singleton.instance;
        this.timestamp = Date.now();
        Singleton.instance = this;
    }
}

const a = new Singleton();
const b = new Singleton();
console.log(a === b); // true


/* -------------------------------------------------------------------------- */
/*                                 factory pattern                            */
/* -------------------------------------------------------------------------- */

class Car { drive() { console.log("Driving a car"); } pen() { console.log("pen")} }
class Truck { drive() { console.log("Driving a truck"); } }

class VehicleFactory {
  static getVehicle(type) {
    if (type === "car") return new Car();
    if (type === "truck") return new Truck();
  }
}

const vehicle = VehicleFactory.getVehicle("car");
vehicle.pen(); // Driving a car
